fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Ordering.java

```diff
@@ -457,60 +546,108 @@ public static <K> PartialOrder<K> mergePartialOrderOfOrderings(@Nonnull Iterable
+                                          final boolean isDistinct) {
+        final var orderingSets = orderings.stream()
+                .map(Ordering::getOrderingSet)
+                .collect(ImmutableList.toImmutableList());
 
-        final var mergedPartialOrder =
-                mergePartialOrderOfOrderings(
-                        () -> orderings
-                                .stream()
-                                .map(Ordering::toPartialOrder)
-                                .iterator());
+        final var mergedOrderingSet = mergePartialOrderOfOrderings(orderingSets);
+
+        final var mergedEqualityBoundKeys =
+                combineEqualityBoundKeys(orderings, combineFn);
 
-        return satisfyingKeyPartsOrdering(mergedPartialOrder, requestedOrdering.getOrderingKeyParts(), Function.identity());
+        return Ordering.ofUnnormalized(mergedEqualityBoundKeys, mergedOrderingSet, isDistinct);
     }
 
     /**
      * Helper method to concatenate the ordering key parts of the participating orderings in iteration order.
      * @param orderings a collection of orderings
+     * @param combineFn a combine function to combine two maps of equality-bound keys (and their bindings)
+     * @return a new ordering representing a concatenation of the given left and right ordering
+     */
+    @Nonnull
+    public static Ordering concatOrderings(@Nonnull final Collection<Ordering> orderings,
+                                           @Nonnull final BinaryOperator<SetMultimap<Value, Comparison>> combineFn) {
+
+        return orderings.stream()
+                .reduce((left, right) -> concatOrderings(left, right, combineFn))
+                .orElseThrow(() -> new RecordCoreException("unable to concatenate orderings"));
+    }
+
+    /**
+     * Helper method to concatenate the ordering key parts of the participating orderings in iteration order.
+     * @param leftOrdering an {@link Ordering}
+     * @param rightOrdering another {@link Ordering} to be concatenated to {@code leftOrdering}
+     * @param combineFn a combine function to combine two maps of equality-bound keys (and their bindings)
      * @return a list of {@link KeyPart}s
      */
     @Nonnull
-    public static List<KeyPart> concatOrderingKeys(@Nonnull Collection<Ordering> orderings) {
+    public static Ordering concatOrderings(@Nonnull final Ordering leftOrdering,
+                                           @Nonnull final Ordering rightOrdering,
+                                           @Nonnull final BinaryOperator<SetMultimap<Value, Comparison>> combineFn) {
+        final var leftOrderingSet = leftOrdering.getOrderingSet();
+        final var rightOrderingSet = rightOrdering.getOrderingSet();
+
+        Verify.verify(leftOrdering.isDistinct());
+        Debugger.sanityCheck(() ->
+                Verify.verify(
+                        Sets.intersection(leftOrderingSet.getSet(), rightOrderingSet.getSet()).isEmpty()));
```

Comment:
would it make sense to enable concatenating posets that share items, or I suppose you're being extra cautious here because that might not be applicable in all cases (because of e.g. introducing cycles)

e.g.
```
PO1: (a -> b)
PO2: (b -> c)
CONCAT(PO1, PO2): (a -> b -> c)
```